src/test/cypress/integration/ProgrammingExerciseStaticCodeAnalysis.spec.ts

```diff
@@ -0,0 +1,76 @@
+import scaSubmission from '../fixtures/programming_exercise_submissions/static_code_analysis/submission.json';
+import { artemis } from '../support/ArtemisTesting';
+import { makeSubmissionAndVerifyResults, startParticipationInProgrammingExercise } from '../support/pageobjects/OnlineEditorPage';
+
+// The user management object
+const users = artemis.users;
+
+// Requests
+const courseManagement = artemis.requests.courseManagement;
+
+// PageObjects
+const editorPage = artemis.pageobjects.programmingExercise.editor;
+const scaConfig = artemis.pageobjects.programmingExercise.scaConfiguration;
+const scaFeedback = artemis.pageobjects.programmingExercise.scaFeedback;
+
+describe('Static code analysis tests', () => {
+    let course: any;
+    let exercise: any;
+
+    before(() => {
+        setupCourseAndProgrammingExercise();
+    });
+
+    it('Configure every static code analysis category to influence grading', () => {
+        cy.login(users.getAdmin());
+        scaConfig.visit(course.id, exercise.id);
+        scaConfig.makeEveryScaCategoryInfluenceGrading();
+        scaConfig.saveChanges();
+    });
+
+    it('Makes successful submission with SCA errors', function () {
+        startParticipationInProgrammingExercise(course.title, exercise.title, users.getStudentOne());
+        makeSuccessfulSubmissionWithScaErrors();
+    });
```

Comment:
Yeah the idea was that the first test adjust the SCA configuration as tutor and then the second test participates as a student and verifies that he gets the SCA feedback and reduced points etc. I agree that it is not great to have tests depending on each other, but there aren't too many options we have. We can either merge all tests, which depend on each other, into one large test or we repeat the setup for every test. One huge test isn't great, because it's not really readable (in this case it would be, but look at the ModelingExercise.spec for example). Repeating the setup  for every sub-test takes a lot of implementation and execution resources.
So I went with this structure, because the tests clearly separate what is tested and are readable

Reply:
I see. The issue I have with it is that it relies on the specific execution order and that the tests are run in the same thread. In case we want to parallelise this in any way or make it more flexible, issues would arise.

Also from looking at a test case, it is not clear that they are depending on each other => Moving something could get messy.

I agree that having blob tests is not a solution and that implementation and execution resources are issues to tackle.

The first two are solvable by extracting certain parts into functions that get called by tests. This also reduces code duplication and you already do it. So I don't see the issue of applying this here as well if the case arises more than just this time.

Execution resources could be an issue but I can't see it in this specific case. Do relate to other test cases where this issue would arise?
If there is a real issue with that, one could extract all dependant tests into a test file and specify the setup in the beforeEach() method.

What do you think?

Reply:
Yeah, I can put it into the same test and refactor the individual steps into their own high-level methods. With execution resources I mostly meant how long the test is running. We need to wait one minute between creating a course and creating a programming exercise. Duplicating the setup for each tests would mean that we probably add about 90 seconds of time to the test for nothing. This could still be avoided if we create only one course and then two programming exercises, but the programming exercise creation itself creates quite some time, so we would probably still add about 20 seconds of execution time.
Furthermore the tests then kinda miss the correlation from the UI interaction in the first test (configuring the SCA grading)  with the effect it has on the second test (that all SCA categories are enabled and reduce points).
So in the first test we would test that we can change the grading configuration in the UI (but don't assert if it actually has an impact on the student exercise).
Moreover I would have to implement an additional request for the second test (so a programming exercise is created, cypress then configures SCA via this request and then the student can participate in the exercise). This request will probably not be used anywhere else and needs to be maintained.
So yeah I prefer moving the tests into one larger test in this case.
If we add parallelisation at some point via the Cypress Dashboard it will[ only parallelize complete spec files](https://docs.cypress.io/guides/guides/parallelization#Overview) not the individual tests inside a spec file.
![image](https://user-images.githubusercontent.com/12861668/133484694-844e189a-0c05-43c4-842c-51c349b54e51.png)

Reply:
Putting it in one test sounds good. Especially since you can split it into multiple (non-test-)functions. But then it is for a new face easy to understand I think.